### DISCUSSION TITLE
buildSubgraphSchema with Apollo federation and authChecker?

### DISCUSSION THREAD

**Discussion by k-atama** [MichalLytek/type-graphql#1240] [Q&A]
I am using Apollo federation with type-graphql. This works, so far, by building the schema using `type-graphql`s `buildSchema`, and then passing the build schema into the `buildSubgraphSchema`, as well as an set of custom resolvers.

```typescript
...

const schema = await buildSchema({
    resolvers,
    container: Container,
    // authChecker doesn't seem to be called :/ 
    authChecker: simpleAuthChecker,
  });

  const server = new Server({
    schema: buildSubgraphSchema({
      typeDefs: gql(printSchema(schema)),
      resolvers: createResolversMap(schema),
    }),
    plugins,
    context: ({ req }) => {
      const context = {
        req,
        user: req.user, // `req.user` comes from `express-jwt`
      };
      return context;
    },
  });

...
```

The above _mostly_ works, however when I'm attempting to add authorization to my project, using the `@Authorization` decorator / process from the [type-graphql documentation](https://typegraphql.com/docs/authorization.html), I'm hitting a road block: it doesn't seem like my authChecker is actually called. 

I haven't dug too deep into the `buildSchema` function yet, I am wondering if I can somehow make this auth function available in the service container or something... 

Anyway, hoping somebody might have an idea on how I can resolve this. If not, hopefully I'm able to find a workaround by stepping through buildSchema or something and I'll report back :). Thank you!

**Comment by MichalLytek** [comment]
`gql(printSchema(schema))` for sure is loosing some features like `@Extensions` because they are not expressible in SDL.
However, I'm not sure why auth is not working for you as it's stored in the metadata storage

**Reply by k-atama** [reply]
Thanks so much for the reply @MichalLytek (and all of your work on the project)! 

I'm assuming this (wanting to use type-graphql as part of a federated API) isn't totally unique, do you have any suggestions on how best to use something like `buildSubgraphSchema` without loosing features (e.g. `@Extensions`, but also I suspect middleware in general maybe doesn't work any longer as I'm using `buildSubgraphSchema` with `gql(printSchema(schema))` currently :/ ).

Thanks again for everything!

**Reply by MichalLytek** [reply]
Something different must be wrong with your setup. Middlewares should work normally.

**Reply by k-atama** [reply]
Thanks again for the reply - so far I haven't been able to track this down, so I'll switch to the non-federated version of the Apollo server to see if I have the same issue. If so, I guess the issue is pehraps something I've done on the config side (though I've stepped through the type-graphql logic on bootstrap and as far as I can tell things are "correct" in terms of how I've configured the custom auth.

One thing I'm not sure is "where" the `authChecker` within the schema after I run (`type-graphql`'s `buildSchema`). E.g.

```
const schema = await buildSchema({
    resolvers,
    // register the 3rd party IOC container
    container: Container,
    authChecker: simpleAuthChecker,
  });
 ```
 
In other words, can I "see" the `authChecker` within the `schema` object above to confirm that at least this portion of the app is configured correctly?

**Reply by MichalLytek** [reply]
Just make the `authChecker` always return `false` so you will see that it's working. Or put `console.log` in the function body.

**Reply by k-atama** [reply]
I've certainly done that part - and unfortunately the console log isn't being hit :/ - this is why I think I have to go a bit deeper in the debugging. 

As mentioned, `type-graphql`'s checks for an `authChecker` on startup from `buildSchema`, and when I include the `authChecker: simpleAuthChecker` then the app starts up fine... it just seems to have no effect unfortunately (the console.log isn't hit, and the route isn't protected). 

When you said "metadata storage", is this something I can see in the computed schema?

**Reply by MichalLytek** [reply]
You can log that in the `global` object, but metadata does not mean runtime schema, it's built from collected metadata
